Improve Yugabyte load balancing setup

krishna-yb opened the pull request:
Improves YugabyteDB load balancing setup for the Sequelize example

- Adds automatic database creation helper (ensure-database.js) that runs before npm start
- Uses standard PostgreSQL environment variables (PGHOST, PGPORT, PGUSER, etc.)
- Adds support for YugabyteDB smart driver features (load balancing, topology awareness) via PG* environment variables
- Configures sequelize-yugabytedb to use @yugabytedb/pg smart driver for automatic load balancing across cluster nodes
- Updates README with comprehensive examples for environment-based configuration
- Sets default port to 5433 (standard YugabyteDB port) with PGPORT override support for testing
- Simplifies configuration to work seamlessly with RF3 YugabyteDB clusters

krishna-yb committed:
Improve Yugabyte load balancing setup

krishna-yb commented:
This is where the env variables are being set in [@yugabytedb/pg](https://www.npmjs.com/package/@yugabytedb/pg?activeTab=readme) package:
It is being set using the **val** function
[Github line link](https://github.com/yugabyte/node-postgres/blob/03e65bcc8ed65267bba2b06e260725bbc1137fe8/packages/pg/lib/connection-parameters.js#L9)


cc: @HarshDaryani896 @ashetkar

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,196 @@
+ * Load balancing configuration via environment variables:
+ * - YB_HOSTS: Comma-separated list of host:port (e.g., "127.0.0.1:5436,127.0.0.2:5436")
+ * - YB_LOAD_BALANCE: Read load balancing mode (any, only-primary, prefer-primary, prefer-rr, only-rr)
+ * - YB_WRITE_LOAD_BALANCE: Write load balancing mode (default: only-primary)
```
```suggestion
 * - YB_LOAD_BALANCE_WRITES_ON: Write load balancing mode (default: only-primary)
```

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,196 @@
+// - 'only-primary': Only uses primary nodes (for writes)
+// - 'prefer-primary': Prefers primary, falls back to RR
+const READ_LOAD_BALANCE_MODE = "any";
+const WRITE_LOAD_BALANCE_MODE = "only-primary";
```
```suggestion
const LOAD_BALANCE_READS_ON = "any";
const LOAD_BALANCE_WRITES_ON = "only-primary";
```

ashetkar reviewed (approved):
LGTM with few comments

krishna-yb committed:
Migrate Sequelize example to use standard PostgreSQL environment variables

Replace YB_* environment variables with standard PostgreSQL PG* variables
for better compatibility and consistency with PostgreSQL tooling.

Changes:
- Replace YB_HOSTS with PGHOSTS (custom multi-host) and PGHOST (standard)
- Replace YB_LOAD_BALANCE with PGLOADBALANCE
- Replace YB_WRITE_LOAD_BALANCE with PGWRITELOADBALANCE
- Replace YB_TOPOLOGY_KEYS with PGTOPOLOGYKEYS
- Replace YB_SERVERS_REFRESH_INTERVAL with PGYBSERVERSREFRESHINTERVAL
- Replace YB_FAILED_HOST_RECONNECT_DELAY_SECS with PGFAILEDHOSTRECONNECTDELAYSECS
- Add support for standard PGUSER, PGPASSWORD, PGDATABASE, PGPORT

New features:
- Add createSequelizeWithConnectionString() for connection string approach
- Support both PGHOSTS (comma-separated multi-host) and PGHOST (single host)
- Environment variables now take precedence over config.json values
- Remove redundant ensureEnv() function
- Update config.json defaults to YugabyteDB standard (yugabyte/yugabyte, port 5436)

Improved documentation:
- Add comprehensive comments explaining variable precedence
- Document all load balancing modes and their behaviors
- Add examples for topology awareness configuration

These changes maintain full backward compatibility while aligning with
PostgreSQL standards and making the configuration more intuitive.

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,232 @@
+ * 
+ * 3. YugabyteDB load balancing variables (extensions to standard PostgreSQL):
+ *    - PGLOADBALANCE: Read load balancing mode (any, only-primary, prefer-primary, prefer-rr, only-rr)
+ *    - PGWRITELOADBALANCE: Write load balancing mode (default: only-primary)
```
We should mention that it is used for the loadBalance property itself, but for writes. And that it is not a standard YB env variable.

ashetkar commented on node/sequelize/README.md:
````diff
@@ -26,11 +26,17 @@ $ DEBUG=sequelize:* npm start
+Environment examples:
+```
+export YB_HOSTS=127.0.0.1:5436,127.0.0.2:5436,127.0.0.3:5436
+export YB_LOAD_BALANCE=any
````
Both are not being used now.

ashetkar commented on node/sequelize/ensure-database.js:
```diff
@@ -0,0 +1,107 @@
+const env = process.env.NODE_ENV || 'development';
+const baseConfig = require(path.join(__dirname, 'config', 'config.json'))[env];
+
+// Parse PGHOSTS (custom, comma-separated) or PGHOST (standard, single host) or use config defaults
```
Why do want to define the custom PGHOSTS env variable for this example?
Also, same functions are duplicated in index.js

krishna-yb committed:
Address PR review comments

- Clarify PGWRITELOADBALANCE documentation (not a standard YB env var, custom extension)
- Update README.md to remove obsolete YB_* variables, add comprehensive PG* examples
- Simplify ensure-database.js to use only standard PGHOST/PGPORT (removed custom PGHOSTS)
- Remove code duplication by eliminating parseHosts from ensure-database.js
- Enhance PGHOSTS documentation explaining why custom variable is needed

krishna-yb committed:
Remove custom PGHOSTS environment variable

- Use only standard PostgreSQL variables (PGHOST, PGPORT, etc.)
- Multi-host load balancing now configured via config.json host field
- Simplified documentation and removed all PGHOSTS references
- config.json can specify comma-separated hosts: '127.0.0.1:5436,127.0.0.2:5436,...'
- Environment variable PGHOST takes precedence over config.json

krishna-yb committed:
Fix load balancing by adding ensureEnv() function

- Added ensureEnv() function to programmatically set environment variables
- @yugabytedb/pg driver reads config from process.env, not Sequelize config
- Simplified Sequelize setup to use single connection (not replication mode)
- Updated config.json to use all 3 nodes for multi-host load balancing
- Updated ensure-database.js to parse multi-host config correctly
- Tested and verified load balancing is working across all 3 nodes

krishna-yb committed:
Clean up code: simplify comments and remove unused variables

- Simplified load balance mode comments
- Removed unused WRITE_LOAD_BALANCE_MODE variable
- Condensed redundant comments in ensureEnv section
- Cleaned up function documentation
- Improved clarity of driver configuration comments
- Kept both connection methods (object config and connection string)

ashetkar commented on node/sequelize/README.md:
```diff
@@ -26,11 +26,41 @@ $ DEBUG=sequelize:* npm start
+
+# Load balancing configuration (YugabyteDB extensions)
+export PGLOADBALANCE=any
+export PGWRITELOADBALANCE=only-primary
```
Not related to any load balance property, needs to be removed.

ashetkar commented on node/sequelize/README.md:
```diff
@@ -26,11 +26,41 @@ $ DEBUG=sequelize:* npm start
+export PGWRITELOADBALANCE=only-primary
+
+# Optional: Topology awareness
+export PGTOPOLOGYKEYS=cloud.region.zone
```
Need to specify with actual values, else remove.

ashetkar commented on node/sequelize/README.md:
````diff
@@ -26,11 +26,41 @@ $ DEBUG=sequelize:* npm start
+export PGTOPOLOGYKEYS=cloud.region.zone
+```
+
+For multi-host load balancing, configure the `host` field in `config/config.json`:
````
What do you mean by multi-host load balancing? We do not use that term while talking about the feature.

ashetkar commented on node/sequelize/README.md:
```diff
@@ -26,11 +26,41 @@ $ DEBUG=sequelize:* npm start
+    "username": "yugabyte",
+    "password": "yugabyte",
+    "database": "ysql_sequelize",
+    "port": "5436"
```
Let's keep the port at default value 5433 everywhere.

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,216 @@
+ * 2. Multi-host configuration via config.json:
+ *    - For YugabyteDB multi-node load balancing, specify multiple hosts in config.json:
+ *      "host": "127.0.0.1:5436,127.0.0.2:5436,127.0.0.3:5436"
+ *    - Environment variable PGHOST takes precedence over config.json
```
Remove

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,216 @@
+ *    - PGFALLBACKTOTOPOLOGYKEYSONLY: Fallback to topology keys only (default: false)
+ *    - PGYBSERVERSREFRESHINTERVAL: Metadata refresh interval in seconds (default: 5)
+ *    - PGFAILEDHOSTRECONNECTDELAYSECS: Reconnect delay in seconds (default: 5)
+ */
```
```suggestion
 * 3. Environment variables for the load balancing properties:
 *    - PGLOADBALANCE: Enable load balancing (Valid values: any, only-primary, prefer-primary, prefer-rr, only-rr)
 *    - PGTOPOLOGYKEYS: Specify the node placements to target (regions/zones) (e.g., "cloud.region.zone")
 *    - PGFALLBACKTOTOPOLOGYKEYSONLY: Fallback to topology keys only (default: false)
 *    - PGYBSERVERSREFRESHINTERVAL: Metadata refresh interval in seconds (default: 5)
 *    - PGFAILEDHOSTRECONNECTDELAYSECS: Delay in reconnecting to failed hosts, in seconds (default: 5)
 */
```

ashetkar commented on node/sequelize/models/index.js:
```diff
@@ -1,29 +1,216 @@
+    (TOPOLOGY_KEYS ? `&topologyKeys=${encodeURIComponent(TOPOLOGY_KEYS)}` : '') +
+    `&ybServersRefreshInterval=${SERVER_REFRESH_INTERVAL}` +
+    `&failedHostReconnectDelaySecs=${FAILED_HOST_RECONNECT_DELAY_SECS}` +
+    (FALLBACK_TO_TOPOLOGY_KEYS_ONLY ? `&fallbackToTopologyKeysOnly=true` : '');
```
Let's not use all that we offer. Users can find that from the smart drivers page.
```suggestion
  const connectionString = 
    `postgres://${encodeURIComponent(username)}:${encodeURIComponent(password)}@${host}:${port}/${database}` +
    `?loadBalance=${READ_LOAD_BALANCE_MODE}` +
    (TOPOLOGY_KEYS ? `&topologyKeys=${encodeURIComponent(TOPOLOGY_KEYS)}` : '') +
    `&ybServersRefreshInterval=${SERVER_REFRESH_INTERVAL}` : '');
```

krishna-yb committed:
Address all review comments and use sequelize-yugabytedb with smart driver

Review comment fixes:
- Remove PGWRITELOADBALANCE (not an official env variable)
- Remove PGTOPOLOGYKEYS example from README
- Remove multi-host terminology and config.json section from README
- Change all ports from 5436 to 5433 (default YugabyteDB port)
- Update env variables documentation wording per suggestions
- Simplify connection string example (removed failedHostReconnectDelaySecs)

Technical changes:
- Use local sequelize-yugabytedb with pg override to @yugabytedb/pg
- sequelize-yugabytedb now handles the pg -> @yugabytedb/pg mapping
- Users no longer need to manually override pg in their package.json
- Add ensureEnv() to ensure-database.js for load balancing support
- Simplified and cleaned up documentation

All changes use only official YugabyteDB environment variables.

HarshDaryani896 commented on node/sequelize/package.json:
```diff
@@ -3,17 +3,20 @@
-    "pg": "^8.6.0",
-    "sequelize-yugabytedb": "^1.0.5",
+    "dotenv": "^16.4.7",
+    "sequelize-yugabytedb": "file:../../sequelize-yugabytedb",
```
@krishna-yb Are there any code changes in our sequelize-yugabytedb package?

krishna-yb committed:
Fix PGPORT environment variable handling to override config ports

- Updated ensure-database.js to respect PGPORT env var even when using config.json
- Updated models/index.js to override all host ports when PGPORT is set
- Ensures PGPORT takes precedence over ports in config.json for testing flexibility

HarshDaryani896 reviewed (approved):
LGTM
